model/fv_dynamics.F90

```diff
@@ -330,15 +333,39 @@ subroutine fv_dynamics(npx, npy, npz, nq_tot,  ng, bdt, consv_te, fill,
       jed = bd%jed
 
 
-!     cv_air =  cp_air - rdgas
-      agrav = 1. / grav
-        dt2 = 0.5*bdt
+!      cv_air =  cp_air - rdgas
+      dt2 = 0.5*bdt
       k_split = flagstruct%k_split
       recip_k_split=1./real(k_split)
       nwat = flagstruct%nwat
       nq = nq_tot - flagstruct%dnats
       nr = nq_tot - flagstruct%dnrts
-      rdg = -rdgas * agrav
+
+      grav_var_h = grav
+      grav_var = grav
+
+      if(flagstruct%var_grav)then
+        do j=js,je
```

Comment:
This loop can be refactored for better locality and cache utilization. I would recommend computing the surface (npz+1) values in the first loop, then compute from npz to 1 by -1 using a  new k-j-i loop. This would also allow you to define `newrad` as a 2D array, reducing memory usage and possibly better fitting in cache.